src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComInterfaceGenerator.cs

```diff
@@ -0,0 +1,338 @@
+{
+    [Generator]
+    public sealed class ComInterfaceGenerator : IIncrementalGenerator
+    {
+        public static class StepNames
+        {
+            public const string CalculateStubInformation = nameof(CalculateStubInformation);
+            public const string GenerateManagedToNativeStub = nameof(GenerateManagedToNativeStub);
+            public const string GenerateNativeToManagedStub = nameof(GenerateNativeToManagedStub);
+        }
+
+        private static readonly ContainingSyntax NativeTypeContainingSyntax = new(
+                                    TokenList(Token(SyntaxKind.InternalKeyword), Token(SyntaxKind.PartialKeyword)),
+                                    SyntaxKind.InterfaceDeclaration,
+                                    Identifier("Native"),
+                                    null);
+
+        public void Initialize(IncrementalGeneratorInitializationContext context)
+        {
+            // Get all methods with the [GeneratedComInterface] attribute.
+            var attributedMethods = context.SyntaxProvider
+                .ForAttributeWithMetadataName(
+                    TypeNames.GeneratedComInterfaceAttribute,
+                    static (node, ct) => node is InterfaceDeclarationSyntax,
+                    static (context, ct) => context.TargetSymbol is INamedTypeSymbol methodSymbol
+                        ? new { Syntax = (InterfaceDeclarationSyntax)context.TargetNode, Symbol = methodSymbol }
+                        : null)
+                .Where(
+                    static modelData => modelData is not null);
+
+            var interfacesWithDiagnostics = attributedMethods.Select(static (data, ct) =>
+            {
+                Diagnostic? diagnostic = GetDiagnosticIfInvalidTypeForGeneration(data.Syntax, data.Symbol);
+                return new { data.Syntax, data.Symbol, Diagnostic = diagnostic };
+            });
+
+            // Split the methods we want to generate and the ones we don't into two separate groups.
+            var interfacesToGenerate = interfacesWithDiagnostics.Where(static data => data.Diagnostic is null);
+            var invalidTypeDiagnostics = interfacesWithDiagnostics.Where(static data => data.Diagnostic is not null);
+
+            context.RegisterSourceOutput(invalidTypeDiagnostics, static (context, invalidMethod) =>
+            {
+                context.ReportDiagnostic(invalidMethod.Diagnostic);
+            });
+
+            var methodsWithDiagnostics = interfacesToGenerate
+                .SelectMany(static (data, ct) =>
+            {
+                ContainingSyntaxContext containingSyntax = new(data.Syntax);
+                Location interfaceLocation = data.Syntax.GetLocation();
+                var methods = ImmutableArray.CreateBuilder<(MethodDeclarationSyntax Syntax, IMethodSymbol Symbol, int Index, Diagnostic? Diagnostic)>();
+                // Start at offset 3 as 0-2 are IUnknown.
+                int methodVtableOffset = 3;
```

Comment:
Is this one of those things we could encapsulate in a "context" for the generator? I could imagine a recursive strategy where we append interfaces and having a "start from index" might make sense. Unsure if that follows with the current design but the callout of IUnknown slots made me think of that.

Reply:
I've split this into a separate context where we can put data that is valid for the whole interface (instead of for individual methods).